### PR TITLE
Switch forward-backward algorithm to log-space for numerical stability

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
## Summary

Refactored the forward-backward algorithm and EM implementation to compute probabilities in log-space, preventing numerical underflow issues with long sequences or many states.

### Core Algorithm
- **Forward-backward algorithm**: Now operates entirely in log-space
  - `forward_pass`: Returns `log_alphas` and `log_norm`
  - `backward_pass`: Returns `log_betas`
  - `forward_backward`: Accepts `log_initial_prob` and `log_transition_prob` parameters
  - Compute joint posteriors (`log_xis`) in log-space using `compute_xi_log`

### EM Algorithm
- **M-step interface**: `run_m_step` now returns log-probabilities for initial and transition matrices
  - Analytical M-step functions still compute in probability space (natural for normalization) - TODO: switch to log-proba, see below.
  - Conversion to log-space happens at the interface boundary - TODO: use only log-proba
- **EM step**: Simplified to directly use log-probabilities from M-step without conversion
- **State management**: `GLMHMMState` stores `log_initial_prob` and `log_transition_matrix`

### Bug Fixes
- Fixed critical bug in `_em_step` where probabilities from M-step were incorrectly stored as log-probabilities
- Updated all test calls to `forward_backward` to pass log-probabilities
- Fixed invalid `is_log` parameter in pytree test

### Benefits
- **Numerical stability**: Prevents underflow with long sequences
- **More accurate posteriors** for models with many states
- **Better convergence** in EM algorithm

### Costs
- Slight efficiency cost. logexpsums operation are slower than dot products.
- For typical use cases (3-10 states), this overhead is negligible

## TODOs

~~- Analytical M-steps via log-probabilities: add complexity, slower, but more robust.~~
~~- Numerical M-step: aggregate_sample_scores using log-sum-exp trick:~~

$$
\cancel{\mathbb{E}_q[\text{nll}] = L^+ - L^-}
$$

where:

$$
\cancel{ L^+ = \exp\left(\text{LSE}\left(\{\log q_i + \log(\text{nll}_i) \mid \text{nll}_i > 0\}\right)\right)}
$$

$$
\cancel{ L^- = \exp\left(\text{LSE}\left(\{\log q_i + \log(-\text{nll}_i) \mid \text{nll}_i < 0\}\right)\right)}
$$

$$
\cancel{ \mathrm{LSE}(x_i) = \log \left(\sum_i e^{x_i}\right)} 
$$

~~Where~~ $\cancel{q}$ ~~is~~ $\cancel{p(z | X)}$ ~~computed in the E-step, and~~ $\cancel{ \log (\text{nll})}$ ~~is ~~$\cancel{-\log p(y | z) }$~~, the log-likelihood of the observation given the states.~~

> The M-step has been kept in proba space, because it looks numerically stable once the posterior are normalized correctly. This both simplifies the code, and makes it more efficient.

### CODE REVIEW

File: src/nemos/glm_hmm/expectation_maximization.py [hunk 1]
```diff
@@ -31,315 +31,360 @@ def _analytical_m_step_initial_prob(
...
+    using alternative optimization methods.
     """
-    tmp_initial_prob = jnp.sum(posteriors, axis=0, where=is_new_session[:, jnp.newaxis])
+    # Mask and sum
+    new_initial_prob = jnp.sum(posteriors, axis=0, where=is_new_session[:, jnp.newaxis])
+
+    # Add prior
     if dirichlet_prior_alphas is not None:
-        tmp_initial_prob += dirichlet_prior_alphas - 1
+        new_initial_prob = new_initial_prob + (dirichlet_prior_alphas - 1)
```
Comment (sjvenditto) [inline src/nemos/glm_hmm/expectation_maximization.py]:
```suggestion
        new_initial_prob += dirichlet_prior_alphas - 1
```

**Review by sjvenditto — APPROVED** [review]
Looks good!
